File: engine/include/xjmusic/entities/meme/ParseStrong.java
// Copyright (c) XJ Music Inc. (https://xj.io) All Rights Reserved.

package io.xj.model.meme;

import java.util.std::vector;
import java.util.Objects;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 Meme Matcher for Strong-Memes
 <p>
 Parse any meme to test if it's valid, and extract its features
 <p>
 Strong-meme like LEMONS! should always favor LEMONS https://github.com/xjmusic/workstation/issues/218
 */
class ParseStrong {
  static final Pattern rgx = Pattern.compile("^(.+)!$");
  final std::string body;
  final boolean isValid;

  ParseStrong(std::string raw) {
    Matcher matcher = rgx.matcher(raw);

    if (!matcher.find()) {
      body = raw;
      isValid = false;
      return;
    }

    body = matcher.group(1);
    if (java.util.Objects.isNull(body) || body.length() == 0) {
      isValid = false;
      return;
    }

    isValid = true;
  }

  public static ParseStrong fromString(std::string raw) {
    return new ParseStrong(raw);
  }

  public boolean isAllowed(std::vector<ParseStrong> memes) {
    if (!isValid) return true;
    for (var meme : memes)
      if (Objects.equals(body, meme.body)) return true;
    return false;
  }
}

File: engine/src/entities/meme/ParseStrong.cpp
// Copyright (c) XJ Music Inc. (https://xj.io) All Rights Reserved.

#include "xjmusic/entities/meme/ParseStrong.h"

using namespace XJ;

ParseStrong::ParseStrong(const std::string &raw) {
  std::smatch matcher;
  isValid = std::regex_search(raw, matcher, rgx);

  if (!isValid) {
    body = raw;
    isValid = false;
    return;
  }

  body = matcher[1].str();
  if (body.empty()) {
    isValid = false;
    return;
  }

  isValid = true;
}

ParseStrong ParseStrong::fromString(const std::string &raw) {
  return ParseStrong(raw);
}

bool ParseStrong::isAllowed(const std::vector<ParseStrong> &memes) {
  if (!isValid) return true;
  return !std::all_of(memes.begin(), memes.end(), [this](const ParseStrong &meme) {
    return body != meme.body;
  });
}
